dulwich/line_ending.py

```diff
@@ -0,0 +1,197 @@
+
+There are sources that seems to indicate that Git won't do line-ending
+normalization when a file contains mixed line-endings. I think this logic
+might be in text / binary detection heuristic but couldn't find it yet.
+
+Sources:
+- https://git-scm.com/docs/git-config#git-config-coreeol
+- https://git-scm.com/docs/git-config#git-config-coreautocrlf
+- https://git-scm.com/docs/gitattributes#_checking_out_and_checking_in
+- https://adaptivepatchwork.com/2012/03/01/mind-the-end-of-your-line/
+"""
+from dulwich.objects import Blob
+
+CRLF = b"\r\n"
+LF = b"\n"
+
+
+def convert_crlf_to_lf(text_hunk):
+    """Convert CRLF in text hunk into LF"""
+    return text_hunk.replace(CRLF, LF)
+
+
+def convert_lf_to_crlf(text_hunk):
+    """Convert LF in text hunk into CRLF"""
+    # TODO find a more efficient way of doing it
+    intermediary = text_hunk.replace(CRLF, LF)
+    return intermediary.replace(LF, CRLF)
+
+
+class NormalizedBlob(Blob):
+    """A Git Blob object with potential line-ending normalization."""
+
+    def __init__(self, blob, conversion, binary_detection):
+        super(NormalizedBlob, self).__init__()
+        self._original_blob = blob
+        self.conversion = conversion
+        self.binary_detection = binary_detection
+        self._needs_serialization = True
+
+    def _serialize(self):
+        original_chunked_text = self._original_blob._chunked_text
+
+        if self.binary_detection is True:
+            # TODO: Support binary detection, in the case where the blob is
+            # detected as binary, no conversion should be done
+            raise NotImplementedError()
+        else:
+            new_chunked_text = map(self.conversion, original_chunked_text)
+
+        return new_chunked_text
+
+
+def get_checkout_filter(core_eol, core_autocrlf, git_attributes):
+    """ Returns the correct checkout filter based on the passed arguments
+    """
+    # TODO this function should process the git_attributes for the path and if
+    # the text attribute is not defined, fallback on the
+    # get_checkout_filter_autocrlf function with the autocrlf value
+    return get_checkout_filter_autocrlf(core_autocrlf)
+
+
+def get_checkin_filter(core_eol, core_autocrlf, git_attributes):
+    """ Returns the correct checkin filter based on the passed arguments
+    """
+    # TODO this function should process the git_attributes for the path and if
+    # the text attribute is not defined, fallback on the
+    # get_checkin_filter_autocrlf function with the autocrlf value
+    return get_checkin_filter_autocrlf(core_autocrlf)
+
+
+def get_checkout_filter_autocrlf(core_autocrlf):
+    """ Returns the correct checkout filter base on autocrlf value
+    """
+
+    if core_autocrlf is True:
+        return convert_lf_to_crlf
+
+    return None
+
+
+def get_checkin_filter_autocrlf(core_autocrlf):
+    """ Returns the correct checkin filter base on autocrlf value
+    """
+
+    if core_autocrlf is True or core_autocrlf == b"input":
```

Comment:
It seems odd that this either takes a boolean or a string. Perhaps just pass in a string, even if that is "true" or "input" ?